src/state/contract_class_cache.rs

```diff
@@ -0,0 +1,140 @@
+//! # Contract cache system
+//!
+//! The contract caches allow the application to keep some contracts within itself, providing them
+//! efficiently when they are needed.
+//!
+//! The trait `ContractClassCache` provides methods for retrieving and inserting elements into the
+//! cache. It also contains a method to extend the shared cache from an iterator so that it can be
+//! used with the private caches.
+//!
+//! TODO: Right now, it's impossible to implement the issue's `merge_caches` because
+//!   `ContractClassCache::extend` will be called already with the write lock in place. To solve
+//!   this, the lock may be pushed into the cache, but the methods will not be able to have
+//!   `&mut self` as the object.
+
+use crate::{services::api::contract_classes::compiled_class::CompiledClass, utils::ClassHash};
+use std::{collections::HashMap, sync::RwLock};
+
+/// The contract class cache trait, which must be implemented by all caches.
+pub trait ContractClassCache {
+    /// Provides the stored contract class associated with a specific class hash, or `None` if not
+    /// present.
+    fn get_contract_class(&self, class_hash: ClassHash) -> Option<CompiledClass>;
+    /// Inserts or replaces a contract class associated with a specific class hash.
+    fn set_contract_class(&self, class_hash: ClassHash, compiled_class: CompiledClass);
+
+    /// Performs a bulk insert of contract classes from an iterator over pairs of the class hash and
+    /// its contract class.
+    fn extend<I>(&self, other: I)
+    where
+        I: IntoIterator<Item = (ClassHash, CompiledClass)>;
+}
+
+// pub(crate) fn merge_caches(
+//     shared: &RwLock<HashMap<ClassHash, CompiledClass>>,
+//     mut private: HashMap<ClassHash, CompiledClass>,
+// ) {
+//     // FIXME: Parallel invocation of `merge_caches()` may cause data loss. Example:
+//     //   - Thread A: Invokes `merge_caches()`. Starts copying data from `shared` into its `private`.
+//     //   - Thread B: Invokes `merge_caches()`. Starts copying data from `shared` into its `private`.
+//     //   - Thread A: Updates the shared cache. It now contains A's specific entries.
+//     //   - Thread B: Updates the shared cache. It now contains only B's specific entries, since it
+//     //     lost A's ones.
```

Comment:
Maybe we can explore using something like [dashmap](https://crates.io/crates/dashmap) instead? Or bite the bullet and modify in-place with the write lock taken, so the same instance is shared instead of swapped.

Reply:
Right now we're extendin the storage with the write lock taken. I don't think it should have any effect on the performance but as it is coded now it shouldn't be impossible to test.

The problem right now is that `ContractClassCache::extend()` accepts an iterator, not a `HashMap<ClassHash, CompiledClass>` because the underlying implementation may not use a `HashMap<ClassHash, CompiledClass>` at all. Making it so that the merge happens without the lock taken would require allocating a temporary container from scratch.